executor/batch_point_get.go

```diff
@@ -449,13 +463,41 @@ func (getter *PessimisticLockCacheGetter) Get(_ context.Context, key kv.Key) ([]
 	return nil, kv.ErrNotExist
 }
 
-func getPhysID(tblInfo *model.TableInfo, intVal int64) int64 {
+func getPhysID(tblInfo *model.TableInfo, partitionExpr *tables.PartitionExpr, intVal int64) (int64, error) {
 	pi := tblInfo.GetPartitionInfo()
 	if pi == nil {
-		return tblInfo.ID
+		return tblInfo.ID, nil
+	}
+
+	if partitionExpr == nil {
+		return tblInfo.ID, nil
 	}
-	partIdx := math.Abs(intVal % int64(pi.Num))
-	return pi.Definitions[partIdx].ID
+
+	switch pi.Type {
+	case model.PartitionTypeHash:
+		partIdx := math.Abs(intVal % int64(pi.Num))
+		return pi.Definitions[partIdx].ID, nil
+	case model.PartitionTypeRange:
+		// we've check the type assertions in func TryFastPlan
+		col := partitionExpr.Expr.(*expression.Column)
```

Comment:
For safety, please check the type again here in case of returning the wrong results.